### PR TITLE
Add support for WHERE clauses to "upsert" queries

### LINKED ISSUE #4145: Add support for WHERE clause to upsert
# Environment

Knex version: 0.12.12
Database + version: Postgres 10.7
OS: MacOS Catalina 10.15.7

# Feature discussion / request

1. Explain what is your use case

I am using the new "upsert" feature added in #3763, which works great! But using PostgreSQL, I need to add a WHERE clause to an "upsert" statement (ON CONFLICT ...), so that I can conditionally "upsert" only certain rows. 

I looked through the Github issues and through the #3763 PR which added support for "upsert" and it doesn't seem like this feature is currently supported.

The PostgreSQL documentation shows the syntax for using a WHERE clause in INSERT ... ON CONFLICT queries: https://www.postgresql.org/docs/10/sql-insert.html.

An example of how this feature is used: https://stackoverflow.com/questions/44571870/postgresql-on-conflict-with-a-where-clause

2. Explain what kind of feature would support this

Adding support for a WHERE clause to "upsert" statements (ON CONFLICT ...) in DB engines that support it would meet my needs.

3. Give some API proposal, how the feature should work

The PostgreSQL compiler (and perhaps MySQL/SQLite compilers?) needs to be updated to support a `.where` method after an `onConflict().merge()` chain.

For PostgreSQL at least, I think the changes would go somewhere around here:
https://github.com/knex/knex/blob/master/lib/dialects/postgres/query/compiler.js#L29

### PR DESCRIPTION
**Pull request by markdboyd**
This is a follow-up to #3763, which adds support for WHERE clauses to "upsert" queries for the DB engines that support this functionality.

Based on my research on the DB engines that support "upsert" queries, it seems like PostgreSQL and SQLite support WHERE clauses for them, but MySQL does not:

https://www.postgresql.org/docs/10/sql-insert.html
https://sqlite.org/lang_upsert.html
https://dev.mysql.com/doc/refman/8.0/en/insert-on-duplicate.html

To do:

- [x] Add support for WHERE clause to upsert queries in PostgreSQL and add tests
- [x] Add support for WHERE clause to upsert queries in SQLite and add tests
- [x] Update documentation: https://github.com/knex/documentation/pull/300

Closes #4145

### CODE REVIEW

File: lib/dialects/postgres/query/compiler.js
Comment (kibertoad) [inline lib/dialects/postgres/query/compiler.js:32]: I would simlify this to a `if (wheres)` check instead of empty concatenation, same as here:
`if (returning) sql += this._returning(returning);`

File: test/integration/builder/inserts.js
Comment (kibertoad) [inline test/integration/builder/inserts.js:1504]: Skipping test altogether same as is done for `redshift` probably would be more transparent in test results log.

File: test/integration/builder/inserts.js
Comment (kibertoad) [inline test/integration/builder/inserts.js:1469]: shouldn't mysql be also skipped if you are saying it is not supported?

**Comment by kibertoad** [comment]
Since you are still planning to add support for SQLite, that will resolve the currently failing test. Could you please add the sanity check for MySQL, which would throw an error if someone tries to use it with WHERE clause, and adjust assertion accordingly?

File: lib/dialects/postgres/query/compiler.js
Comment (markdboyd) [reply lib/dialects/postgres/query/compiler.js:32]: Will do

File: test/integration/builder/inserts.js
Comment (markdboyd) [reply test/integration/builder/inserts.js:1504]: I don't fully understand the test setup - but it seems like the difference between redshift and oracle/mssql is that for oracle/mssql we expect an error to be thrown, but not redshift? Should I preserve that distinction or just add mssql/oracle to the skip?

File: lib/dialects/postgres/query/compiler.js
Comment (markdboyd) [reply lib/dialects/postgres/query/compiler.js]: 706ce6e42513c52547c045827a504a32991e941b

**Comment by markdboyd** [comment]
@kibertoad I think I have addressed your feedback, minus one question I had about the integration tests

File: lib/dialects/mysql/query/compiler.js
Comment (markdboyd) [inline lib/dialects/mysql/query/compiler.js:32]: It feels like there might be a better way to check whether `where` is called after a `merge` statement for MySQL, but I could not figure out what it is

File: test/integration/builder/inserts.js
Comment (kibertoad) [reply test/integration/builder/inserts.js:1509]: TBH, I think we don't even run redshift in CI, so feel free to just keep ignoring it.
BTW, shouldn't MySQL also be included in error assertion?

Comment (markdboyd) [reply test/integration/builder/inserts.js:1509]: Ah yeah. Added MySQL to have an error assertion:

90cba493009f09ceb4457c4040cfb1da406129d2

File: test/integration/builder/inserts.js
Comment (kibertoad) [inline test/integration/builder/inserts.js:1469]: since you are asserting the error now, probably no longer you need to skip it

Comment (markdboyd) [reply test/integration/builder/inserts.js]: Sorry about that. Fixed: 80f366d67fd90d4792470a3418668849c54daa36

**Comment by kibertoad** [comment]
This can be merged after documentation is added to https://github.com/knex/documentation

**Comment by markdboyd** [comment]
@kibertoad OK, I have created a documentation PR: https://github.com/knex/documentation/pull/300

**Review by kibertoad — APPROVED** [review]

**Comment by kibertoad** [comment]
Thanks a lot!

**Comment by kibertoad** [comment]
Released in 0.21.13

**Pull request merged**
